server/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/list/TransportListTasksAction.java

```diff
@@ -66,19 +81,61 @@ protected void taskOperation(Task actionTask, ListTasksRequest request, Task tas
     }
 
     @Override
-    protected void processTasks(ListTasksRequest request, Consumer<Task> operation) {
+    protected void processTasks(ListTasksRequest request, Consumer<Task> operation, Runnable nodeOperation, Consumer<Exception> onFailure) {
         if (request.getWaitForCompletion()) {
-            long timeoutNanos = waitForCompletionTimeout(request.getTimeout());
-            operation = operation.andThen(task -> {
+            Set<Task> matchedTasks = Sets.newConcurrentHashSet();
+            CountDownLatch matchedTasksReady = new CountDownLatch(1);
+            AtomicBoolean nodeOperationFiredOff = new AtomicBoolean();
+            // Register the listener before we start iterating over tasks and sync it with a latch,
+            // because matched tasks can get removed whilst we are iterating over them
+            var listener = new RemovedTaskListener() {
+                @Override
+                public void onRemoved(Task task) {
+                    try {
+                        matchedTasksReady.await();
+                    } catch (InterruptedException e) {
+                        Thread.currentThread().interrupt();
```

Comment:
Is this aim here to bubble up the a new InterruptedException? I guess if this happens, then the timeout exception scheduled operation below will also fire up ultimately?

Reply:
I don't think we should be blocking here. Could we instead remember the prematurely-removed tasks captured here and process them when we're ready to do so?

Reply:
Something like this perhaps?

```java
ListenableActionFuture<Void> future = new ListenableActionFuture<>();
Set<Task> removedTasks = Sets.newConcurrentHashSet();
Set<Task> matchedTasks = Sets.newConcurrentHashSet();
RefCounted removalRefs = AbstractRefCounted.of(() -> {
    for (Task removedTask : removedTasks) {
        matchedTasks.remove(removedTask);
    }
    removedTasks.clear();
    if (matchedTasks.isEmpty()) {
        future.onResponse(null);
    }
});

RemovedTaskListener removedTaskListener = task -> {
    if (removalRefs.tryIncRef()) {
        removedTasks.add(task);
        removalRefs.decRef();
    } else {
        matchedTasks.remove(task);
        if (matchedTasks.isEmpty()) {
            future.onResponse(null);
        }
    }
};

taskManager.addRemovedTaskListener(removedTaskListener);
try {
    processTasks(request, task -> {
        if (task.getAction().startsWith(ListTasksAction.NAME) == false) {
            // It doesn't make sense to wait for List Tasks and it can cause an infinite loop of the task waiting
            // for itself or one of its child tasks
            matchedTasks.add(task);
        }
        operation.accept(task);
    });
} finally {
    removalRefs.decRef();
}

final ActionListener<Void> listener = ActionListener.runBefore(
    ActionListener.wrap(ignored -> nodeOperation.run(), onFailure),
    () -> taskManager.removeRemovedTaskListener(removedTaskListener)
);

if (future.isDone()) {
    listener.onResponse(null);
} else {
    future.addListener(new ThreadedActionListener<>(logger, threadPool, ThreadPool.Names.MANAGEMENT, listener, false));
    threadPool.schedule(
        () -> future.onFailure(new ElasticsearchTimeoutException("Timed out waiting for completion of tasks")),
        requireNonNullElse(request.getTimeout(), DEFAULT_WAIT_FOR_COMPLETION_TIMEOUT),
        ThreadPool.Names.SAME
    );
}
```